NEW Improved file used on table data

emteknetnz opened the pull request:
Issue https://github.com/silverstripe/silverstripe-admin/issues/1098

Blocked by https://github.com/silverstripe/silverstripe-framework/pull/9735

maxime-rainville commented:
What's the expected behaviour when an object references itself? Right now it puts two entries in the table.

 ![image](https://user-images.githubusercontent.com/1168676/96421807-b4cf4480-1253-11eb-8307-672a2fde83a4.png)

emteknetnz commented:
@maxime-rainville `File::class` is excluded from showing on an [asset-admin extension](https://github.com/silverstripe/silverstripe-asset-admin/pull/1139/files#diff-42aa76520722b7a3235caf3db312e40d66f7b1fa4def1c231135eed227200095R20), which I put there to exclude Folder::class from showing because Folder doesn't have its own database table and the RelatedDataService class uses database tables instead of the ORM.

This really does seems like a bit of and edge case though

maxime-rainville commented:
Self-referential objects probably won't happen very often, but there's plenty of scenarios were it makes perfect sense. e.g.: A business is using its own product ... so they track themselves a supplier.

It doesn't sound like something very difficult to fix either.

maxime-rainville reviewed (changes requested):
The `<tbody>` tag should have a `aria-live="polite"` attribute.

We've got a proper [loading component](https://silverstripe.github.io/silverstripe-admin/?selectedKind=Admin%2FLoading&selectedStory=Loading&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybook%2Fnotes%2Fpanel).

Should we take steps to hide records the current user doesn't have read access on?

emteknetnz commented on README.md:
```diff
@@ -37,3 +37,45 @@ All methods, with `public` visibility, are part of the public API. All other met
 
 Please [create an issue](http://github.com/silverstripe/silverstripe-admin/issues) for any bugs you've found, or features you're missing.
+
+## Customising the File "Used on" table in the Files section (asset-admin)
```
docs.silverstripe.org?  I don't believe that's the correct place to document module specific things?  If it is and I've missed something, could you please show me where it's supposed to go alongside other module specific docs?

chillu replied:
doc.silverstripe.org *is* the correct place for core module specific docs, otherwise we fragment the hell out of the information. Stuff that's not in core (e.g. `silverstripe/blog`) should be documented on the repo, but that's simply because we haven't had the time to pull all of this into doc.silverstripe.org. In an ideal world, doc.silverstripe.org would be your starting point for anything in supported modules, but that's hard to create a consistent IA around.

emteknetnz replied:
I can't find a relevant section on docs.silverstripe.org to put information about implementing an extension for an admin feature.  Where should it go?

maxime-rainville replied:
Maybe as a separate article under https://docs.silverstripe.org/en/4/developer_guides/files/ ?

emteknetnz commented on client/src/components/UsedOnTable/provideUsedOnData.js:
```diff
@@ -66,8 +85,12 @@ const provideUsedOnData = (UsedOnTable) => {
 
-  const connectedUsedOnDataProvider = connect(mapStateToProps, { loadUsedOn })(UsedOnDataProvider);
-  connectedUsedOnDataProvider.Component = UsedOnDataProvider;
+  const ComponentWithTabContext = injectTabContext(UsedOnDataProvider);
+  const connectedUsedOnDataProvider = connect(
+    mapStateToProps,
+    { loadUsedOn }
+  )(ComponentWithTabContext);
+  connectedUsedOnDataProvider.Component = ComponentWithTabContext;
 
   return connectedUsedOnDataProvider;
```
Doing it that way fails a unit test
```
Summary of all failing tests
 FAIL  client/src/components/UsedOnTable/tests/UsedOnTable-test.js
  ● UsedOnTable › provideUsedOnData › Component behaviour › should try to load data onMount

    Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: undefined. You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.

maxime-rainville replied:
Humm ... not sure why this is set up this way. My instinct would be to scrap this and refactor the entire thing, but that's probably another job for another day.

emteknetnz commented on README.md:
```diff
@@ -37,3 +37,45 @@ All methods, with `public` visibility, are part of the public API. All other met
+
+Extension hooks can be used to do the following:
+a) Exclude DataObjects of a particular type of class from showing in results
+b) Link ancestors of a DataObject so they show on the same row of results
```
Updated

emteknetnz commented on README.md:
````diff
@@ -37,3 +37,45 @@ All methods, with `public` visibility, are part of the public API. All other met
+Example PHP file:
+
+```php
+namespace My\App\Extensions;
````
Updated

emteknetnz commented on README.md:
```diff
@@ -37,3 +37,45 @@ All methods, with `public` visibility, are part of the public API. All other met
+a) Exclude DataObjects of a particular type of class from showing in results
+b) Link ancestors of a DataObject so they show on the same row of results
+
+Example PHP file:
```
Updated

emteknetnz commented on client/src/components/Tabs/tests/TabNav-test.js:
```diff
@@ -7,10 +7,10 @@ import Adapter from 'enzyme-adapter-react-16/build/index';
 
 Enzyme.configure({ adapter: new Adapter() });
 
+const onToggle = jest.fn();
```
Updated

emteknetnz commented on client/src/components/Tabs/tests/Tabs-test.js:
```diff
@@ -7,6 +7,8 @@ import Adapter from 'enzyme-adapter-react-16/build/index';
 
 Enzyme.configure({ adapter: new Adapter() });
 
+const activateTab = jest.fn();
```
Updated

emteknetnz commented on code/Forms/UsedOnTable.php:
```diff
@@ -66,39 +67,51 @@ public function getSchemaDataDefaults()
      * @param HTTPRequest $request
      * @return HTTPResponse $response
+     *
+     * @throws \ReflectionException
```
It was something to do with ClassInfo in the framework pull-request, have fixed it over on that side and removed the `@throws` from the docblock

emteknetnz commented on code/Forms/UsedOnTable.php:
```diff
@@ -66,39 +67,51 @@ public function getSchemaDataDefaults()
-        $response = new HTTPResponse(json_encode($data));
-        return $response
-            ->addHeader('Content-Type', 'application/json');
+        $response = new HTTPResponse(json_encode(['usage' => $usageData]));
```
Updated

emteknetnz commented on README.md:
```diff
@@ -37,3 +37,45 @@ All methods, with `public` visibility, are part of the public API. All other met
+    }
+}
+
+Example YML file:
```
Updated

emteknetnz commented on client/src/components/UsedOnTable/UsedOnTable.js:
```diff
@@ -11,7 +11,6 @@ class UsedOnTable extends PureComponent {
         <tr>
           <th className="used-on__col--index">#</th>
```
Updated

emteknetnz commented:
>> The <tbody> tag should have a aria-live="polite" attribute.

Updated

>> We've got a proper loading component.

Added

>> Should we take steps to hide records the current user doesn't have read access on?

What constitutes read access? canView() only works on the front-end, not in the CMS.  For example adding the following code to Page.php means the Page will still be visible in the CMS, however you won't be able to view it in the frontend:

**Page.php**
```php
public function canView($member = null) {
    if ($this->Title == 'Contact Us') return false;
    return parent::canView($member);
}
```

chillu commented on client/src/components/UsedOnTable/UsedOnTable.js:
```diff
@@ -44,7 +44,7 @@ class UsedOnTable extends PureComponent {
 
       return (
-        <tbody>
+        <tbody aria-live="polite">
```
Nice, thank you for caring about those details Steve!

maxime-rainville reviewed (changes requested):
I'm not a big fan of this error message. When something goes wrong, can we catch the error and display something a little more friendly to our beloved users?

![image](https://user-images.githubusercontent.com/1168676/96945031-00763c80-1539-11eb-8ddb-438d4fa3a329.png)

Obvious errors we might want to plan for:
- the file has been deleted;
- your session has expired;
- you don't have access to view the file.

You make a good point on the `canView` issue. Maybe have a chat with PUX about what they want to see there, but I don't have concerns with the current behaviour security wise.

emteknetnz commented:
@maxime-rainville 
>> When something goes wrong, can we catch the error and display something a little more friendly to our beloved users

That error message for "something went wrong" is existing behaviour and feels like scope creep on a long running feature.  We'd need to do a bunch of work to identity the different types of potential issues, write units tests etc, probably a 5 point story in itself.  I'd say split this off as a seperate task if we want to improve it.

I'd be more concerned with how you managed to trigger that error rather than the messaging - any idea how you managed to do that?

emteknetnz commented on code/Forms/UsedOnTable.php:
```diff
@@ -70,35 +71,53 @@ public function getSchemaDataDefaults()
     public function usage(HTTPRequest $request)
     {
         $record = $this->getRecord();
```
I don't think we need to, I've manually tested that users without permission to the file won't be able to see file in the file manager.

If we're concerned about an authenticated cms user make POST requests to see usage on a file they don't have view permissions on (pretty unlikely scenario):

UsedOnTable extends FormField

$this->getRecord() will call $this->getForm()->getRecord()

Seems like the appropriate place to add protection would be somewhere higher up on the form than the used on table formfield

maxime-rainville replied:
We should aim to have a multi layered approach the security. Assumptions like "access has already been validated by this other method" is how you introduce security vulnerability.

You've got a method whose sole purpose is to handle a REST request. That's a natural place to do a existence and read-access check.

emteknetnz replied:
I've added a `$record->canView()` check to the `usage()` function

sachajudd reviewed (changes requested):
I'm unsure what's going on with using links in the Used on table. I'm not sure as to why we even have links when links are not in the designs? A user will not be taken to other pages separately from where the file is. They will be taken to where that file is being used directly as this tab intends to give that information e.g:  'New blocks page / A picture' = the whole row is clickable and you are taken to 'A picture' which opens up the content block as to where that particular file lives in the CMS. No links are needed to other pages. Right now this is *very* inaccessible and messy. A user wouldn't know that a 'title' is a link unless they were, by chance to hover over it. Other pages I've tested there are no links at all and no way to get to the file even though I can hover on the row? There's no consistency here. This seems to have gone down its own path of design without informing @silverstripeux of these changes. 
![image](https://user-images.githubusercontent.com/24258161/97250927-d63fba00-186b-11eb-97d5-dc7c323917f0.png)

Rows that can be hovered but can't be clicked, also where are your cursor: hover styles?
![image](https://user-images.githubusercontent.com/24258161/97244582-f4062280-185d-11eb-8dd9-7c9cc68afea0.png)

You need to look at the designs more closely. The numbers are not meant to be in bold. The 'type' should be 11px, different from the 'title' which is 13px. I've added the Invision link to [DSM](https://projects.invisionapp.com/dsm/silver-stripe/silverstripe-cms/nav/build/asset/componentContainers/5f152068f6772b33db2e52b0/tab/design?mode=edit).

The number should align with the title. I can see it's slightly off here probably because of the border:
![image](https://user-images.githubusercontent.com/24258161/97248144-fb312e80-1865-11eb-8279-31b625addc15.png)

If the image title is too long you can truncate the text at a certain width:
![image](https://user-images.githubusercontent.com/24258161/97246926-6f1e0780-1863-11eb-9b39-bec174efdef1.png)

You could use an alert for the error messages. That is a standard pattern we use.

emteknetnz commented:
@sachajudd I've made the following changes:
- Removed the individual links
- Made the whole row clickable, it'll go to the linked DataObject, or the nearest ancestor with a CMS edit link -so in the case of content blocks that lack a CMS edit link it will go the to page edit form
- Titles that are 35 characters or more are truncated to 32 characters followed by `...`
- Other styles adjustments such as font sizes, removing bold, etc

![image](https://user-images.githubusercontent.com/4809037/97267443-0e0c2900-188f-11eb-965c-ed40909dc45a.png)

I've update UAT shortly

sachajudd reviewed (changes requested):
Following up a conversation from @clarkepaul here: https://github.com/silverstripeltd/product-issues/issues/249#issuecomment-681484971

> If we end up including files from form submissions they should link to the submission and not the form.

On UAT this is not happening at the moment. The 'type' also reads 'Submitted File Field' I'm not sure how I feel about this wording here. Since this is describing the type e.g Content block, Page, Table data or Campaign, I think it should just read 'Form Submission'.

Something funky going on here with the padding in the table row:
![image](https://user-images.githubusercontent.com/24258161/97364363-d4780400-1908-11eb-8e7a-3fe344077345.png)

The type here should read 'File block'. Not just File, I'm guessing this pulls directly from the list but it would be good to add 'block' onto the end if possible.
![image](https://user-images.githubusercontent.com/24258161/97364578-23259e00-1909-11eb-9948-ff7ce022dcf6.png)

emteknetnz commented:
Outcome from team meeting this morning was that things are generally looking good.  The solution of clicking a row instead of links and going to the related DataObject, or if not CMSLink, the nearest ancestors edit form, was agreed to be pretty good.

I should be able to implement the rest of the design feedback

maxime-rainville reviewed (approved):
I'm happy where this is from a technical point of view.

sachajudd reviewed (changes requested):
Few quick things:
* Can we change “This is not used anywhere” to "This file is currently not in use."
*  Truncation seems to not be currently working
* Linking to the form submission tab also seems to not be working

After we've fixed those few things I'm happy to approve from a UX perspective 🙂

emteknetnz commented:
First two points should be easy to fix up, I can do that shortly

Last point about linking to form submissions, this one's more complex so I'll need to raise a new issue for it

sachajudd commented:
Let's stick with the current behaviour with truncation as we talked about offline and I'll create a new issue with some designs to solve the few issues we are having. # DSM link To do

sachajudd reviewed (approved):
Approved from a UX perspective. Nice work Steve this was a huge effort to get across the line! 🎉 🙂 

I will follow up and add a few UX issues and we can chat further on the other more technical PR's that are to be prioritised.

michalkleiner commented:
Would the wording
> Can we change “This is not used anywhere” to "This file is currently not in use."

still work for the use case where a file is uploaded and published on the website just to have a shareable URL for it been considered? It is still used from the CMS editor's perspective, it's just assigned to any relation to another CMS object.

Perhaps tweak to "This file is currently not in use within the CMS" or similar?

brynwhyman commented:
IMO I think the change from the existing wording to "This file is currently not in use." is a good change and gives enough context that the file hasn't been added/ _used on something_. 

A change to make the wording more specific to use within the CMS still isn't going to help the editor know whether the file's URL has been added to content.
